container.go

```diff
@@ -0,0 +1,595 @@
+}
+
+// Shutdown requests a container shutdown, but it may not actually be shutdown until Wait() succeeds.
+func (container *container) Shutdown() error {
+	operation := "Shutdown"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	var (
+		//options string
+		resultp *uint16
+	)
+
+	err := hcsShutdownComputeSystemTP5(container.handle, nil, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		if err == ErrVmcomputeOperationPending {
+			return ErrVmcomputeOperationPending
+		}
+		err = &ContainerError{Container: container, Operation: operation, Err: err}
+		logrus.Error(err)
+		return err
+	}
+
+	logrus.Debugf(title+" succeeded id=%s", container.id)
+	return nil
+}
+
+// Terminate requests a container terminate, but it may not actually be terminated until Wait() succeeds.
+func (container *container) Terminate() error {
+	operation := "Terminate"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	var (
+		//options string
+		resultp *uint16
+	)
+
+	err := hcsTerminateComputeSystemTP5(container.handle, nil, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		if err == ErrVmcomputeOperationPending {
+			return ErrVmcomputeOperationPending
+		}
+		err = &ContainerError{Container: container, Operation: operation, Err: err}
+		logrus.Error(err)
+		return err
+	}
+
+	logrus.Debugf(title+" succeeded id=%s", container.id)
+	return nil
+}
+
+// Wait synchronously waits for the container to shutdown or terminate.
+func (container *container) Wait() error {
+	operation := "Wait"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	if hcsCallbacksSupported {
+		callbackNumber, err := container.registerCallback(hcsNotificationSystemExited)
+		if err != nil {
+			return &ContainerError{Container: container, Operation: operation, Err: err}
+		}
+		defer container.unregisterCallback(callbackNumber)
+
+		err = container.waitForNotification(callbackNumber)
+		if err != nil {
+			return &ContainerError{Container: container, Operation: operation, Err: err}
+		}
+
+		logrus.Debugf(title+" succeeded id=%s", container.id)
+		return nil
+	}
+
+	_, err := container.waitTimeoutInternal(syscall.INFINITE)
+	if err != nil {
+		return &ContainerError{Container: container, Operation: operation, Err: err}
+	}
+	return nil
+}
+
+func (container *container) waitTimeoutInternal(timeout uint32) (bool, error) {
+	return waitTimeoutInternalHelper(container, timeout)
+}
+
+// WaitTimeout synchronously waits for the container to terminate or the duration to elapse. Returns
+// ErrTimeout if the operation times out in HCS
+func (container *container) WaitTimeout(timeout time.Duration) error {
+	operation := "WaitTimeout"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	if hcsCallbacksSupported {
+		callbackNumber, err := container.registerCallback(hcsNotificationSystemExited)
```

Comment:
The helper should contain both the register and the wait, so the defer unregister can just happen in the helper